platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,998 @@
+  public final SolrClient getClient() {
+    // returning this to make sure all calls from our SolrClient API will still be intercepted
+    // when the client is retrieved and passed to a Solr request object
+    // this should be temporary until we completely abstract out Solr with interfaces
+    return this;
+  }
+
+  @Override
+  public String getCore() {
+    return core;
+  }
+
+  @Override
+  @SuppressWarnings("squid:S2093" /* closing of real client handled by finalizeStateChange() */)
+  public void close() throws IOException {
+    try {
+      if (state == State.CLOSED) {
+        return;
+      }
+      final SolrClient previousClientToClose;
+      final Future<?> futureToCancel;
+
+      synchronized (lock) {
+        if (state == State.CLOSED) { // already closed so bail
+          return;
+        }
+        futureToCancel = future;
+        previousClientToClose = realClient;
+        LOGGER.debug("Solr({}): closing", core);
+        this.unavailableClient =
+            new UnavailableSolrClient(
+                new UnavailableSolrException("'" + core + "' client was closed"));
+        this.apiClient = unavailableClient;
+        this.pingClient = unavailableClient;
+        this.realClient = null;
+        this.state = State.CLOSED;
+        this.future = null;
+        lock.notifyAll(); // wakeup those waiting for isAvailable(timeout)
+      }
+      finalizeStateChange(true, futureToCancel, previousClientToClose, false);
+    } finally {
+      listeners.clear();
+      initializers.clear();
+    }
+  }
+
+  @Override
+  public boolean isAvailable() {
+    final boolean available = (state == State.CONNECTED);
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(
+          "Solr({}): current availability is [{} = {}]",
+          core,
+          state,
+          SolrClientAdapter.availableToString(available));
+    }
+    if (available && wasNotRecent(lastPing, SolrClientAdapter.PING_MIN_FREQUENCY)) {
+      LOGGER.debug(
+          "Solr({}): Proxy is starting a background task to ping the client because the last ping was too long ago",
+          core);
+      executor.submit(this::backgroundPing);
+    }
+    return available;
+  }
+
+  @Override
+  public boolean isAvailable(long timeout, TimeUnit unit) throws InterruptedException {
+    Validate.notNull(unit, "invalid null time unit");
+    if (isAvailable()) { // quick check to avoid synchronization
+      return true;
+    }
+    // letting now be recomputed by the timedWait() method allows us to better control testing
+    long now = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis());
+    final long end = now + unit.toNanos(timeout);
+
+    synchronized (lock) {
+      while (true) {
+        if (state == State.CLOSED) {
+          return false;
+        }
+        final boolean available = isAvailable();
+
+        if (available) {
+          return true;
+        }
+        final long t = end - now;
+
+        if (t <= 0L) { // we timed out
```

Comment:
I cannot remove `t` as I need it to be computed so I only wait for what is left. Remember that we can be awaken before the end. So since I have to compute `t`, I would rather compute it once and check that value. Comparing `now` and `end` once and then subtracting them after will just yield twice the same thing. Plus, with this approach, `t` representing the time left to sleep. So if it is 0 (or negative), then it is clear that there is no more time left.

Reply:
Just thought that `now >= end` was clearer than `t <= 0L`. Actually, renaming `t` to `timeRemaining` might help here ;-)